README.md
NAME
====

Data::Crumbr - Render data structures for easy searching and parsing

SYNOPSIS
========

    use Data::Crumbr;

    say crumbr(\%options)->($data_structure);


    $ cat sample.json
    {
        "ciao": [
            "a",
            [],
            "tutti",
            {
                "quanti": "voi",
                "hey ☺": {},
                "some": 123.123
            }
        ]
    }

    $ crumbr sample.json
    {"ciao"}[0]:"a"
    {"ciao"}[1]:[]
    {"ciao"}[2]:"tutti"
    {"ciao"}[3]{"hey ☺"}:{}
    {"ciao"}[3]{"quanti"}:"voi"
    {"ciao"}[3]{"some"}:123.123

    $ crumbr --style uri sample.json
    /ciao/0 "a"
    /ciao/1 []
    /ciao/2 "tutti"
    /ciao/3/hey%20%E2%98%BA {}
    /ciao/3/quanti "voi"
    /ciao/3/some 123.123


ALL THE REST
============

Want to know more? [See the module's documentation](http://search.cpan.org/perldoc?App::Crumbr) to figure out
all the bells and whistles of this module!

Want to install the latest release? [Go fetch it on CPAN](http://search.cpan.org/dist/App::Crumbr/).

Want to contribute? [Fork it on GitHub](https://github.com/polettix/App::Crumbr).

That's all folks!



Changes
{{$NEXT}}

Changes for Data::Crumbr:

0.1.1     2015-12-10 10:39:12+01:00 Europe/Rome
   - removed dependency on perl 5.18
   - tidied code a bit

0.1.0     2015-09-28 15:29:53+02:00 Europe/Rome
   - initial release on CPAN


t/data.pl
use utf8;
my $data = {
   one => '1',
   two => 2,
   three => 3.1,
   four => '4.0',
   array => [ qw< what ever >, { inner => 'part', empty => [] } ],
   hash => {
      'with ♜' => {},
      ar => [ 1..3 ],
      something => "funny \x{263A} ☻",
   },
};


lib/Data/Crumbr/Util.pm
package Data::Crumbr::Util;

# ABSTRACT: utility functions for Data::Crumbr
use strict;
use Carp;
use Scalar::Util qw< reftype blessed >;

sub json_leaf_encoder {
   require B;
   return \&_json_leaf_encode;
}

{
   my $slash_escaped;

   BEGIN {
      $slash_escaped = {
         0x22 => '"',
         0x5C => "\\",
         0x2F => '/',
         0x08 => 'b',
         0x0C => 'f',
         0x0A => 'n',
         0x0D => 'r',
         0x09 => 't',
      };
   } ## end BEGIN

   sub _json_leaf_encode {
      return 'null' unless defined $_[0];

      my $reftype = ref($_[0]);
      return '[]' if $reftype eq 'ARRAY';
      return '{}' if $reftype eq 'HASH';
      return (${$_[0]} ? 'true' : 'false')
        if $reftype eq 'SCALAR';

      if (my $package = blessed($_[0])) {
         my $reftype = reftype($_[0]);
         return (${$_[0]} ? 'true' : 'false')
           if ($reftype eq 'SCALAR') && ($package =~ /bool/mxsi);
      }

      croak "unsupported ref type $reftype" if $reftype;

      my $number_flags = B::SVp_IOK() | B::SVp_NOK();
      return $_[0]
        if (B::svref_2object(\$_[0])->FLAGS() & $number_flags)
        && 0 + $_[0] eq $_[0]
        && $_[0] * 0 == 0;

      my $string = join '', map {
         my $cp = ord($_);

         if (exists $slash_escaped->{$cp}) {
            "\\$slash_escaped->{$cp}";
         }
         elsif ($cp >= 32 && $cp < 128) {    # ASCII
            $_;
         }
         elsif ($cp < 0x10000) {             # controls & BML
            sprintf "\\u%4.4X", $cp;
         }
         else {                              # beyond BML
            my $hi = ($cp - 0x10000) / 0x400 + 0xD800;
            my $lo = ($cp - 0x10000) % 0x400 + 0xDC00;
            sprintf "\\u%4.4X\\u%4.4X", $hi, $lo;
         }
      } split //, $_[0];
      return qq<"> . $string . qq<">;
   } ## end sub _json_leaf_encode
}

sub uri_encoder {
   require Encode;
   return \&_uri_encoder;
}

{
   my %is_unreserved;

   BEGIN {
      my @u = ('a' .. 'z', 'A' .. 'Z', '0' .. '9', qw< - _ . ~ >);
      %is_unreserved = map { $_ => 1 } @u;
   }

   sub _uri_encoder {
      my $octets = Encode::encode('UTF-8', $_[0], Encode::FB_CROAK());
      return join '',
        map { $is_unreserved{$_} ? $_ : sprintf('%%%2.2X', ord $_); }
        split //, $octets;
   } ## end sub _uri_encoder
}

sub id_encoder {
   return sub { $_[0] };
}

1;
__END__

=pod

=encoding utf-8

=head1 DESCRIPTION

Utility functions for Data::Crumbr.

=head2 INTERFACE

=over

=item B<< id_encoder >>

   my $encoder = id_encoder();

trivial encoding function that just returns its first argument (i.e. no
real encoding is performed).

=item B<< json_leaf_encoder >>

   my $encoder = json_leaf_encoder();

encoding function that returns a JSON-compliant value, only for leaf
values. It works on:

=over

=item *

plain strings, returned after JSON encoding (e.g. tranformation of
newlines, etc.)

=item *

empty array references, in which case string C<[]> is returned

=item *

empty hash references, in which case string C<{}> is returned

=item *

null values, in which case string C<null> is returned

=back

=item B<< uri_encoder >>

   my $encoder = uri_encoder();

encoding function that then encodes strings according to URI encoding
(i.e. percent-encoding).

=back

=cut


lib/Data/Crumbr/Default.pm
package Data::Crumbr::Default;

# ABSTRACT: Default renderer for Data::Crumbr

use Mo qw< default coerce >;

use strict;
use warnings;
use Carp;
use English qw< -no_match_vars >;
use Scalar::Util qw< blessed >;
use Data::Crumbr::Util;

my $jenc = Data::Crumbr::Util::json_leaf_encoder();
my $ienc = Data::Crumbr::Util::id_encoder();

has array_open        => (default => sub { '' });
has array_close       => (default => sub { '' });
has array_key_prefix  => (default => sub { '[' });
has array_key_suffix  => (default => sub { ']' });
has array_key_encoder => (default => sub { $ienc });
has hash_open         => (default => sub { '' });
has hash_close        => (default => sub { '' });
has hash_key_prefix   => (default => sub { '{' });
has hash_key_suffix   => (default => sub { '}' });
has hash_key_encoder  => (default => sub { $jenc });
has value_encoder     => (default => sub { $jenc });
has keys_separator    => (default => sub { '' });
has value_separator   => (default => sub { ':' });

has output => (
   default => sub { __output() },
   coerce  => \&__output,
);

sub __output {
   my ($output) = @_;
   $output //= [];
   my $reftype = ref $output;

   if (!$reftype) {    # filename, transform into filehandle
      my $fh = \*STDOUT;
      if ($output ne '-') {
         $fh = undef;
         open $fh, '>', $output
           or croak "open('$output'): $OS_ERROR";
      }
      binmode $fh, ':raw'
        or croak "binmode() on $output: $OS_ERROR";
      $reftype = ref($output = $fh);
   } ## end if (!$reftype)

   return sub {
      return unless @_;
      print {$output} $_[0], "\n";
     }
     if $reftype eq 'GLOB';

   return sub {
      return $output unless @_;
      push @$output, $_[0];
     }
     if $reftype eq 'ARRAY';

   return sub {
      return unless @_;
      $output->print($_[0]);
     }
     if blessed($output);

   return sub {
      return unless @_;
      return $output->($_[0]);
     }
     if $reftype eq 'CODE';

   croak "invalid output";
} ## end sub __output

sub leaf {
   my ($self, $stack) = @_;

   my $venc       = $self->value_encoder();
   my @components = $venc->($stack->[-1]{data});

   my @keys = map { $_->{encoded} } @$stack;
   shift @keys;    # first item of @$stack is dummy
   pop @keys;      # last item of @$stack is the leaf, drop it

   my $closers = '';
   if (@keys) {
      unshift @components, join $self->keys_separator(), @keys;
      $closers = $stack->[-2]{closers};
   }

   my $record = join $self->value_separator(), @components;
   $self->output()->($record . $closers);
} ## end sub leaf

{
   no strict 'refs';
   *scalar_leaf = \&leaf;
   *array_leaf  = \&leaf;
   *hash_leaf   = \&leaf;
}

sub array_keys_iterator {
   my ($self, $aref) = @_;
   my $i   = 0;
   my $sup = @$aref;
   return sub {
      return if $i >= $sup;
      return $i++;
   };
} ## end sub array_keys_iterator

sub hash_keys_iterator {
   my ($self, $href) = @_;
   my @keys = sort keys %$href;    # memory intensive...
   return sub { return shift @keys };
}

sub array_key {
   my ($self, $key) = @_;
   return join '', $self->array_open(),
     $self->array_key_prefix(),
     $self->array_key_encoder()->($key),
     $self->array_key_suffix();
} ## end sub array_key

sub hash_key {
   my ($self, $key) = @_;
   return join '', $self->hash_open(),
     $self->hash_key_prefix(),
     $self->hash_key_encoder()->($key),
     $self->hash_key_suffix();
} ## end sub hash_key

sub result {
   my ($self) = @_;
   my $output = $self->output()->()
     or return;
   return join "\n", @$output;
} ## end sub result

sub reset {
   my ($self) = @_;
   my $output = $self->output()->()
     or return;
   @$output = ();
   return;
} ## end sub reset

1;
__END__

=pod

=encoding utf-8

=head1 DESCRIPTION

This is the default encoder implementation, and most probably the only
oney you really need. And most probably, you really not need to directly
use it.

=head1 INTERFACE

=over

=item B<< array_key >>

returns the encoded array key, optionally opening an array and keeping
into account the prefix, the suffix and the encoder for the key

=item B<< array_keys_iterator >>

returns an iterator sub starting from 0 up to the number of elements in
the array

=item B<< hash_key >>

returns the encoded hash key, optionally opening an hash and keeping
into account the prefix, the suffix and the encoder for the key

=item B<< hash_keys_iterator >>

returns an iterator sub that returns each key in the input hash, sorted
lexicographically

=item B<< leaf >>

=item B<< array_leaf >>

=item B<< hash_leaf >>

=item B<< scalar_leaf >>

this method is called whenever an external iteration component hits a
leaf and wants to push a new encoded record to the output

=item B<< new >>

   my $enc = Data::Crumbr::Default->new(%args);

create a new encoder object

=item B<< reset >>

reset the encoder, i.e. wipe out all the internal state to start a new
encoding cycle.

=item B<< result >>

get the outcome of the encoding. Not guaranteed to work.

=back

=cut


lib/Data/Crumbr/Default/URI.pm
package Data::Crumbr::Default::URI;

# ABSTRACT: "JSON" profile for Data::Crumbr::Default
use Data::Crumbr::Util;

sub profile {
   my $json_encoder = Data::Crumbr::Util::json_leaf_encoder();
   my $uri_encoder  = Data::Crumbr::Util::uri_encoder();
   return {
      array_open       => '',
      array_close      => '',
      array_key_prefix => '',
      array_key_suffix => '',

      hash_open       => '',
      hash_close      => '',
      hash_key_prefix => '',
      hash_key_suffix => '',

      keys_separator  => '/',
      value_separator => ' ',

      array_key_encoder => $uri_encoder,
      hash_key_encoder  => $uri_encoder,
      value_encoder     => $json_encoder,
   };
} ## end sub profile

1;
__END__

=pod

=encoding utf-8

=head1 DESCRIPTION

Profile for URI encoder

=head1 INTERFACE

=over

=item B<< profile >>

   my $profile = Data::Crumbr::Default::URI->profile();

returns a default profile, i.e. encoder data to be used to instantiate a
Data::Crumbr::Default encoder. See L</Data::Crumbr> for details about
this profile.

=back

=cut


lib/Data/Crumbr/Default/JSON.pm
package Data::Crumbr::Default::JSON;

# ABSTRACT: "JSON" profile for Data::Crumbr::Default
use Data::Crumbr::Util;

sub profile {
   my $json_encoder = Data::Crumbr::Util::json_leaf_encoder();
   return {
      hash_open       => '{',
      hash_key_prefix => '',
      hash_key_suffix => ':',
      hash_close      => '}',

      array_open       => '[',
      array_key_prefix => '',
      array_key_suffix => '',
      array_close      => ']',

      keys_separator    => '',
      value_separator   => '',
      array_key_encoder => sub { },
      hash_key_encoder  => $json_encoder,
      value_encoder     => $json_encoder,
   };
} ## end sub profile

1;
__END__

=pod

=encoding utf-8

=head1 DESCRIPTION

Profile for JSON encoder

=head1 INTERFACE

=over

=item B<< profile >>

   my $profile = Data::Crumbr::Default::JSON->profile();

returns a default profile, i.e. encoder data to be used to instantiate a
Data::Crumbr::Default encoder. See L</Data::Crumbr> for details about
this profile.

=back

=cut


lib/Data/Crumbr.pm
package Data::Crumbr;

# ABSTRACT: Render data structures for easy searching and parsing

# Inlined Mo
use Mo qw< default coerce >;

use strict;
use warnings;
use Carp;
use English qw< -no_match_vars >;
use Exporter qw< import >;
use Scalar::Util qw< blessed >;

our @EXPORT      = qw< crumbr >;
our @EXPORT_OK   = @EXPORT;
our %EXPORT_TAGS = (all => [@EXPORT_OK]);

has encoder => (
   default => sub { __encoder() },
   coerce  => \&__encoder,
);

sub __load_class {
   my ($class) = @_;
   (my $packname = "$class.pm") =~ s{::}{/}gmxs;
   require $packname;
   return $class;
} ## end sub __load_class

sub crumbr {
   my %args = (@_ && ref($_[0])) ? %{$_[0]} : @_;
   if (defined(my $name = delete $args{profile})) {
      my $class   = __PACKAGE__ . "::Default::$name";
      my $profile = __load_class($class)->profile();
      my $encoder = delete($args{encoder}) // {};
      %$encoder = (
         %$profile,
         %$encoder,    # allow some overriding
         class => '::Default',    # but not on this one
      );
      %args = (encoder => $encoder);
   } ## end if (defined(my $name =...))
   my $wh = __PACKAGE__->new(%args);
   return sub { $wh->encode(@_) };
} ## end sub crumbr

sub __encoder {
   my ($e) = @_;
   if (!blessed($e)) {
      my ($class, @parameters) = $e;
      if (ref($e) eq 'HASH') {
         $class      = delete $e->{class};
         @parameters = %$e;
      }
      $class = '::Default' unless defined $class;
      $class = __PACKAGE__ . $class
        if substr($class, 0, 2) eq '::';
      $e = __load_class($class)->new(@parameters);
   } ## end if (!blessed($e))
   return $e;
} ## end sub __encoder

sub encode {
   my ($self, $data) = @_;
   my $encoder = $self->encoder();
   $encoder->reset();

   my @stack = ({closers => ''}, {data => $data, type => ref($data)},);
 ITERATION:
   while (@stack > 1) {    # frame #0 is dummy
      my $frame = $stack[-1];
      if ($frame->{type} eq 'ARRAY') {
         if (!scalar(@{$frame->{data}})) {
            $encoder->array_leaf(\@stack);
         }
         else {
            my $iterator = $frame->{iterator} //=
              $encoder->array_keys_iterator($frame->{data});
            if (defined(my $key = $iterator->())) {
               $frame->{encoded} = $encoder->array_key($key);
               $frame->{closers} =
                 $encoder->array_close() . $stack[-2]{closers};
               my $child_data = $frame->{data}[$key];
               push @stack,
                 {
                  data => $child_data,
                  type => ref($child_data),
                 };
               next ITERATION;
            } ## end if (defined(my $key = ...))
         } ## end else [ if (!scalar(@{$frame->...}))]
      } ## end if ($frame->{type} eq ...)
      elsif ($frame->{type} eq 'HASH') {
         if (!scalar(keys %{$frame->{data}})) {
            $encoder->hash_leaf(\@stack);
         }
         else {
            my $iterator = $frame->{iterator} //=
              $encoder->hash_keys_iterator($frame->{data});
            if (defined(my $key = $iterator->())) {
               $frame->{encoded} = $encoder->hash_key($key);
               $frame->{closers} =
                 $encoder->hash_close() . $stack[-2]{closers};
               my $child_data = $frame->{data}{$key};
               push @stack,
                 {
                  data => $child_data,
                  type => ref($child_data),
                 };
               next ITERATION;
            } ## end if (defined(my $key = ...))
         } ## end else [ if (!scalar(keys %{$frame...}))]
      } ## end elsif ($frame->{type} eq ...)
      else {    # treat as leaf scalar
         $encoder->scalar_leaf(\@stack);
      }

      # only leaves or end-of-container arrive here
      pop @stack;
   } ## end ITERATION: while (@stack > 1)

   return $encoder->result();
} ## end sub encode

1;

=pod

=encoding utf-8

=head1 SYNOPSIS

   use Data::Crumber; # imports `crumbr`

   # some data to work with
   my $data = { what => 'ever', hey => 'you' };

   # crumbr provides an anonyous sub back. This has defaults
   my $csub = crumbr();

   # use it to encode the data
   my $encoded = $crumbr->($data);
   # {"here"}{"hey"}:"you"
   # {"here"}{"what"}:"ever"

   # URI profile simplifies things but loses something
   $encoded = crumbr(profile => 'URI')->($data);
   # here/hey "you"
   # here/what "ever"

   # JSON profile produces valid JSON "slices"
   $encoded = crumbr(profile => 'JSON')->($data);
   # {"here":{"hey":"you"}}
   # {"here":{"what":"ever"}}

   # Object Oriented Interface
   my $crobj = Data::Crumbr->new();
   $encoded = $crobj->encode($data); # same as default


=head1 DESCRIPTION

Data::Crumbr lets you render data structures in a way that can then be
easily searched and parsed in "slices". The basic idea is that data
shaped in this way will then be easily filtered in the shell for
extracting interesting parts.

The input data structure is traversed is if it is a tree (so no
circular structures please!), and a I<record> is generated for each leaf
in the tree. Depending on the backend and the configurations, the full
path from the root to the parent of the leaf is represented as a
sequence of keys (which can be hash keys or array indexes) followed by
the value. This should make your life easier e.g. in the shell, so that
you can specify the full path to the data structure part you're
interested into with common Unix tools like C<grep> and/or C<sed>.

=head2 Example

Suppose you have the following data structure in Perl:

   my $data = {
      one => '1',
      two => 2,
      three => 3.1,
      four => '4.0',
      true => \1,
      false => \0,
      array => [
         qw< what ever >,
         { inner => 'part', empty => [] }
       ],
      hash => {
         'with ♜' => {},
         ar => [ 1..3 ],
         something => "funny \x{263A} ☻",
      },
   };

If you encode this e.g. in JSON, it will be easy to parse with
the right program, but not from the shell, even if you pretty
print it:

   {
      "hash" : {
         "something" : "funny ☺ ☻",
         "with ♜" : {},
         "ar" : [
            1,
            2,
            3
         ]
      },
      "one" : "1",
      "array" : [
         "what",
         "ever",
         {
            "inner" : "part",
            "empty" : []
         }
      ],
      "four" : "4.0",
      "true" : true,
      "two" : 2,
      "three" : 3.1,
      "false" : false
   }

How do you get the second item in the array C<ar>i inside the
hash C<hash>? Would you do better with YAML instead?

   ---
   array:
     - what
     - ever
     - empty: []
       inner: part
   false: !!perl/ref
     =: 0
   four: 4.0
   hash:
     ar:
       - 1
       - 2
       - 3
     something: funny ☺ ☻
     with ♜: {}
   one: 1
   three: 3.1
   true: !!perl/ref
     =: 1
   two: 2

Not really. Data::Crumbr lets you represent the data in a
more verbose but easily consumable way for the shell. Hence,
this:

   use Data::Crumbr;
   print crumbr()->($data), "\n";

will give you this:

   {"array"}[0]:"what"
   {"array"}[1]:"ever"
   {"array"}[2]{"empty"}:[]
   {"array"}[2]{"inner"}:"part"
   {"false"}:false
   {"four"}:"4.0"
   {"hash"}{"ar"}[0]:1
   {"hash"}{"ar"}[1]:2
   {"hash"}{"ar"}[2]:3
   {"hash"}{"something"}:"funny \u263A \u263B"
   {"hash"}{"with \u265C"}:{}
   {"one"}:"1"
   {"three"}:3.1
   {"true"}:true
   {"two"}:2

Now it should pretty easy for a shell program to get at the
data, e.g. with this C<sed> substitution:

   sed -ne 's/^{"hash"}{"ar"}\[2\]://p'

=head2 Profiles

If you don't like the default encoding, you can get a different
one by using a I<profile>. This is a set of configurations for
C<Data::Crumbr::Default>, which is a pretty generic class for
representing a wide class of possible record-oriented encodings.

A C<Data::Crumbr::Default> encoder is defined in terms of the following
parameters:

=over

=item C<array_open>

sequence to put when an array is opened

=item C<array_close>

sequence to put when an array is closed

=item C<array_key_prefix>

sequence to put before an array's index

=item C<array_key_suffix>

sequence to put after an array's index

=item C<array_key_encoder>

a reference to a function that encodes an array's index

=item C<hash_open>

sequence to put when a hash is opened

=item C<hash_close>

sequence to put when a hash is closed

=item C<hash_key_prefix>

sequence to put before a hash's key

=item C<hash_key_suffix>

sequence to put after a hash's key

=item C<hash_key_encoder>

a reference to a function that encodes a hash's key

=item C<value_encoder>

a reference to a function that encodes a leaf value

=item C<keys_separator>

sequence to separate the keys breadcrumb

=item C<value_separator>

sequence to separate the keys from the value

=back

By default, Data::Crumbr ships with the following profiles:

=over

=item B<< Default >>

i.e. the profile you get by default, and what you saw in action in the
example above. It has the following settings:

=over

=item *

no openers and closers:

   array_open  => ''
   array_close => ''
   hash_open   => ''
   hash_close  => ''

=item *

array keys are printed verbatim, surrounded by square brackets:

   array_key_prefix  => '['
   array_key_suffix  => ']'
   array_key_encoder => Data::Crumbr::Util::id_encoder

=item *

hash keys encoded as JSON strings, surrounded by curly brackets:

   hash_key_prefix   => '['
   hash_key_suffix   => ']'
   hash_key_encoder  => Data::Crumbr::Util::json_leaf_encoder

=item *

no separator between keys (because they already stand out very clearly,
but a colon to separate the sequence of keys from the value:

   keys_separator  => ''
   value_separator => ':'

=item *

leaf values encoded as JSON scalars:

   value_encoder => Data::Crumbr::Util::json_leaf_encoder

=back

This is quite verbose, but lets you specify very precisely what you are
looking for because the hash keys stand out clearly with respect to
array identifiers, i.e. there's no chance that you will mistake an array
index for a hash key (because they are embedded in different bracket
types).

=item B<< JSON >>

this profile always provides you compact JSON-compliant string
representations that contain only one single leaf value.

It has the following characteristics:

=over

=item *

openers and closers are what you would expect for JSON objects and
arrays:

   array_open  => '['
   array_close => ']'
   hash_open   => '{'
   hash_close  => '}'

=item *

there is only one non-empty suffix, i.e. the hash key suffix, so that
we can separate the hash key from the value with C<:> according to JSON:

   array_key_prefix => ''
   array_key_suffix => ''
   hash_key_prefix  => ''
   hash_key_suffix  => ':'

=item *

array keys are not printed:

   array_key_encoder => sub { }

=item *

hash keys are JSON encoded:

   hash_key_encoder  => Data::Crumbr::Util::json_encoder()

=item *

no separators are needed:

   keys_separator  => ''
   value_separator => ''

=item *

leaf values encoded as JSON scalars:

   value_encoder => Data::Crumbr::Util::json_leaf_encoder

=back



=item B<< URI >>

this is the simplest of the profiles, and sacrifices the possibility to
distinguish between hash and array keys to the altar of simplicity.

It has the following characteristics:

=over

=item *

no openers, closers, prefixes or suffixes:

   array_open  => ''
   array_close => ''
   array_key_prefix => ''
   array_key_suffix => ''

   hash_open   => ''
   hash_close  => ''
   hash_key_prefix => ''
   hash_key_suffix => ''

=item *

array keys are printed verbatim

=item *

hash keys are URI encoded

   hash_key_encoder  => Data::Crumbr::Util::uri_encoder

=item *

keys are separated by a slash character C</> and values are separated by
a single space C< >:

   keys_separator  => '/'
   value_separator => ' '

=item *

leaf values encoded as JSON scalars:

   value_encoder => Data::Crumbr::Util::json_leaf_encoder

=back

=back



=head1 INTERFACE

There are two ways to use Data::Crumber: a function C<crumbr>,
that is exported by default, and the object-oriented interface.

=over

=item B<< crumbr >>

   $subref = crumbr(%args); # OR
   $subref = crumbr(\%args);

get a I<crumbr> generator based on provided C<%args>.

Returns a reference to a sub, which can then be called upon a data
structure in order to get the I<crumbed> version.

The input arguments can be:

=over

=item C<< encoder >>

details about the encoder, see L</Profiles> for the available key-value
pairs. In addition, you can also set the following:

=over

=item C<< output >>

the output channel to use for sending encoded data. This can be:

=over

=item * I<filename>

this will be opened in raw mode and used to send the output

=item * I<filehandle>

used directly

=item * I<array reference>

each output line will be pushed as a new element in the array

=item * I<object reference>

which is assumed to support the C<print()> method, that will be called
with each generated line

=item * I<sub reference>

which will be called with each generated line

=back

=back

=item C<< profile >>

the name of a profile to use as a base - see L</Profiles>. Settings in
the profile are always overridden by corresponding ones in the provided
encoder, if any.

=back

=item B<< encode >>

   $dc->encode($data_structure);

generate the encoding for the provided C<$data_structure>. Output is
generated depending on how it is specified, see L</crumbr> above.

=item B<< new >>

   my $dc = Data::Crumber->new(encoder => \%args);

create a new instance of C<Data::Crumbr>. Data provided for the
C<encoder> parameter (i.e. C<%args>) are those discussed in
L</Profiles>.

The new instance can then be used to encode data using the C</encode>
method.
   
=back

=cut

__END__
